cockatrice/src/client/ui/widgets/visual_deck_editor/visual_deck_editor_widget.cpp

```diff
@@ -0,0 +1,347 @@
+    proxyModel->setFilterRole(Qt::DisplayRole);
+
+    completer = new QCompleter(proxyModel, this);
+    completer->setCompletionRole(Qt::DisplayRole);
+    completer->setCompletionMode(QCompleter::PopupCompletion);
+    completer->setCaseSensitivity(Qt::CaseInsensitive);
+    completer->setFilterMode(Qt::MatchContains);
+    completer->setMaxVisibleItems(15);
+    searchBar->setCompleter(completer);
+
+    // Update suggestions dynamically
+    connect(searchBar, &QLineEdit::textEdited, searchModel, &CardSearchModel::updateSearchResults);
+    connect(searchBar, &QLineEdit::textEdited, this, [=, this](const QString &text) {
+        // Ensure substring matching
+        QString pattern = ".*" + QRegularExpression::escape(text) + ".*";
+        proxyModel->setFilterRegularExpression(QRegularExpression(pattern, QRegularExpression::CaseInsensitiveOption));
+
+        if (!text.isEmpty()) {
+            completer->complete(); // Force the dropdown to appear
+        }
+    });
+
+    connect(completer, static_cast<void (QCompleter::*)(const QString &)>(&QCompleter::activated), this,
+            [=, this](const QString &completion) {
+                // Prevent the text from changing automatically when navigating with arrow keys
+                if (searchBar->text() != completion) {
+                    searchBar->setText(completion);                           // Set the completion explicitly
+                    searchBar->setCursorPosition(searchBar->text().length()); // Move cursor to the end
+                }
+            });
+
+    // Ensure that the text stays consistent during selection
+    connect(searchBar, &QLineEdit::textEdited, this, [=, this](const QString &text) {
+        if (searchBar->hasFocus() && !searchBar->completer()->popup()->isVisible()) {
+            // Allow text to change when typing, but not when navigating the completer
+            QString pattern = ".*" + QRegularExpression::escape(text) + ".*";
+            proxyModel->setFilterRegularExpression(
+                QRegularExpression(pattern, QRegularExpression::CaseInsensitiveOption));
+        }
+    });
+
+    // Search button functionality
+    searchPushButton = new QPushButton(this);
+    connect(searchPushButton, &QPushButton::clicked, this, [=, this]() {
+        CardInfoPtr card = CardDatabaseManager::getInstance()->getCard(searchBar->text());
+        if (card) {
+            emit cardAdditionRequested(card);
+        }
+    });
+
+    searchLayout->addWidget(searchBar);
+    searchLayout->addWidget(searchPushButton);
+
+    mainLayout->addWidget(searchContainer);
+
+    groupAndSortContainer = new QWidget(this);
+    groupAndSortLayout = new QHBoxLayout(groupAndSortContainer);
+    groupAndSortLayout->setAlignment(Qt::AlignLeft);
+    groupAndSortContainer->setLayout(groupAndSortLayout);
+
+    groupByComboBox = new QComboBox();
+    QStringList groupProperties;
+    groupProperties << "maintype"
```

Comment:
You use << definition here, but bracket definition below. Maybe we should stay consistent with formatting?